cpp/src/arrow/compute/kernels/nth_to_indices.cc

```diff
@@ -0,0 +1,205 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/kernels/nth_to_indices.h"
+
+#include <algorithm>
+#include <numeric>
+#include <vector>
+
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/expression.h"
+#include "arrow/compute/logical_type.h"
+#include "arrow/type_traits.h"
+
+namespace arrow {
+
+class Array;
+
+namespace compute {
+
+/// \brief Kernel implementing NthToIndices operation
+class ARROW_EXPORT NthToIndicesKernel : public OpKernel {
```

Comment:
By the way, all private code should be enclosed in the anonymous namespace (and you can remove `ARROW_EXPORT`).

Reply:
removed ARROW_EXPORT.
I'm kind of confused about how compute kernels are used by client code.
Exported functions like SortToIndices will create a kernel every time client invokes it. Looks to me exporting the kernel class interface is better, so clients can create a kernel, then call it against different inputs repeatedly.